README.md
# eagle-data-publisher

## Requirements
- **Python 3.8**
    - **FYI**: Change python version; e.g. from `python3.6` into `python3.8`
        - Install python3.8: `$ sudo apt install python3.8`
        - Install venv (optional): `$ apt install python3.8-venv`
        - Install python3.8 dev: `$ sudo apt-get install python3.8-dev`
        - Change alternative: `$ sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 1`
            - It will get response as follows:
                ```
                update-alternatives: using /usr/bin/python3.8 to provide /usr/bin/python3 (python3) in auto mode
                ```
        - Set: `$ sudo update-alternatives --config python3`
            - It will get response as follows:
                ```
                There is only one alternative in link group python3 (providing /usr/bin/python3): /usr/bin/python3.8
                Nothing to configure.
                ```
        - Check result: `$ python3 --version`

## Installation
1. Clone project: `$ git clone https://github.com/ardihikaru/eagle-data-publisher.git`
2. Export pycore (can be done later):
    `export PYTHONPATH=:<root-path-to-project>/eagle-data-publisher/pycore`
    - The path may be different. :)
        - In Eaglestitch Server: 
             `export PYTHONPATH=:/home/eagles/devel/eagle-data-publisher/pycore`
        - In Popeye Server: 
             `export PYTHONPATH=:/home/popeye/devel/eagle-data-publisher/pycore`
        - In LittleBoy Server: 
             `export PYTHONPATH=:/home/s010132/devel/eagle-data-publisher/pycore`
3. Install required library
    - Make sure to use latest version of `pip`:
        `$ pip3 install --upgrade pip`
    - Install `Cython`: `$ pip3 install Cython`
    - Install rest of the library: `$ pip3 install -r requirements.txt`
    - OpenCV, it is better to install them manually.
        - For personal use, you may use easy install by running this: `$ pip3 install opencv-python`
        - Especially for `JetsonNano`, you should **NEVER** try to use **pip install**.
            - Follow this [tutorial to do installation in JetsonNano](https://github.com/ardihikaru/eagle-data-publisher/blob/main/Jetson_Nano_Setup_for_ADS) 
4. Finish. you may follow `Testing usage` or `How to use` section.

## Testing usage
1. Prepared two terminals
2. Run Consumer:
    - Add lib (if not added yet), e.g.: 
        `$ export PYTHONPATH=:/home/ardi/devel/nctu/IBM-Lab/eagle-data-publisher/pycore`
    - Run consumer:
        `$ env RUST_LOG=debug python3 data_consumer.py -l tcp/localhost:7446`
            - If you try them in the same PC, you can use `localhost`
            - If you use to test them to PubSub with different PCs, change them to **IP of the Consumer**
3. Run Publisher
    - Add lib (if not added yet), e.g.: 
        `$ export PYTHONPATH=:/home/ardi/devel/nctu/IBM-Lab/eagle-data-publisher/pycore`
    - Run publisher:
        `$ env RUST_LOG=debug python3 data_publisher.py -e tcp/localhost:7446 --resize`
            - If you try them in the same PC, you can use `localhost`
            - If you use to test them to PubSub with different PCs, change them to **IP of the Consumer**

## How to use
1. Run Consumer service
2. Run Publisher service (**this project**)
    - Sample script to run in the field trial:
        - Popeye:
            - Default: `$ python3 data_publisher.py -e tcp/192.168.1.11:7446 --resize`
            - Resize Camera's resolution: `$ python3 data_publisher.py -e tcp/192.168.1.11:7446 --resize --pwidth 848 --pheight 480`
    - Run sample: `$ python3 data_publisher.py -e tcp/localhost:7446 --cvout --resize`
        - Possible parameters:
            - `--cvout`: Show the real-time captured images; this can be omitted if you run this script with no monitor capability
            - `--resize`: Force to resize the captured image into **FullHD**; can be omitted
            - `-e <path>`: a protocol to transmit data; **Simply focus to change**:
                - `localhost` into the target Consumer's IP Address, and
                - `7446` into any desired port.
            - `-v <path_to_video>`; Path to extract the video stream; 
            if not provided, it load **video0** (value `0`) from your hardware.
            - `--pwidth <target_width> --pheight <target_height>`: 
            Target width & height to change the OpenCV property. 
            The default value is a FullHD (`1920x1080`)
                - If you have no idea, use this command to show possible resolution (**Linux only**):
                `$ v4l2-ctl --list-formats-ext`, and you will receive response similar like this:
                  ``` 
                  $ v4l2-ctl --list-formats-ext                                                                            ^(*features/experiment-2021-06-23-cont+996) 14:17:59 
                    ioctl: VIDIOC_ENUM_FMT
                        Type: Video Capture
                    
                        [0]: 'MJPG' (Motion-JPEG, compressed)
                            ...
                            Size: Discrete 352x288
                                Interval: Discrete 0.033s (30.000 fps)
                            Size: Discrete 640x480
                                Interval: Discrete 0.033s (30.000 fps)
                        [1]: 'YUYV' (YUYV 4:2:2)  <<<--- Focus here. :)
                            Size: Discrete 1280x720
                                Interval: Discrete 0.100s (10.000 fps)
                            ...
                            Size: Discrete 640x480
                                Interval: Discrete 0.033s (30.000 fps)  
                  ```
                  - Please **ONLY Focus on the format** `YUYV`. :)

## MISC
- Change python version; e.g. from `python3.6` into `python3.8`
    - Install python3.8: `$ sudo apt install python3.8`
    - Install venv (optional): `$ apt install python3.8-venv`
    - Install python3.8 dev: `$ sudo apt-get install python3.8-dev`
    - Change alternative: `$ sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 1`
        - It will get response as follows:
            ```
            update-alternatives: using /usr/bin/python3.8 to provide /usr/bin/python3 (python3) in auto mode
            ```
    - Set: `$ sudo update-alternatives --config python3`
        - It will get response as follows:
            ```
            There is only one alternative in link group python3 (providing /usr/bin/python3): /usr/bin/python3.8
            Nothing to configure.
            ```
    - Check result: `$ python3 --version`

CHANGELOG.md
## 0.1.0 (July 01, 2021)
  - init files
  - Initial commit



data_publisher.py
from eagle_zenoh.zenoh_lib.zenoh_net_publisher import ZenohNetPublisher
import sys
import time
from datetime import datetime
import numpy as np
import cv2
import simplejson as json
from enum import Enum
import logging
import argparse
# from hurry.filesize import size as fsizeb
from eagle_zenoh.extras.functions import humanbytes as fsize
from eagle_zenoh.zenoh_lib.functions import encrypt_str, get_img_fsize_in_float
import csv
import os

# PATH TO Save CSV File
CSV_FILE_PATH = "./bandwidth_usage.csv"

# FullHD Format; fixed value, as per required in 5G-DIVE Project
FULLHD_WIDTH = 1920
FULLHD_HEIGHT = 1080

# Encoding parameter
encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), 70]  # The default value for IMWRITE_JPEG_QUALITY is 95

# --- [START] Command line argument parsing --- --- --- --- --- ---
parser = argparse.ArgumentParser(
	description='Zenoh Publisher example')
parser.add_argument('--peer', '-e', dest='peer',  # e.g. `-e tcp/192.168.1.10:7446` (to LittleBoy)
                    metavar='LOCATOR',
                    action='append',
                    type=str,
                    help='Peer locators used to initiate the zenoh session.')
parser.add_argument('--path', '-p', dest='path',
                    default='/eagle/svc/zenoh-python-pub',
                    type=str,
                    help='The name of the resource to publish.')
parser.add_argument('--video', '-v', dest='video',
                    default="0",
                    type=str,
                    help='The name of the resource to publish.')
parser.add_argument('--pwidth', dest='pwidth', default=1920, type=int, help='Target width to publish')
parser.add_argument('--pheight', dest='pheight', default=1080, type=int, help='Target height to publish')
parser.add_argument('--cvout', dest='cvout', action='store_true', help="Use CV Out")
parser.add_argument('--resize', dest='resize', action='store_true', help="Force resize to FullHD")
parser.set_defaults(cvout=False)
parser.set_defaults(resize=False)

args = parser.parse_args()
print(args)
# --- [END] Command line argument parsing --- --- --- --- --- ---

###

L = logging.getLogger(__name__)


###

peer = args.peer
if peer is not None:
	peer = ",".join(args.peer)

video_path = args.video
if video_path == "0":
	video_path = int(video_path)

# Enable / disable cvout
_enable_cv_out = args.cvout

# configure zenoh service
path = args.path
z_svc = ZenohNetPublisher(
	_path=path, _session_type="PUBLISHER", _peer=peer
)
z_svc.init_connection()

# register and collect publisher object
z_svc.register()
publisher = z_svc.get_publisher()

#########################
# Zenoh related variables

window_title = "img-data-publisher"

published_height, published_weight = args.pheight, args.pwidth  # target width and height
cam_weight, cam_height = None, None  # default detected width and height

cap = cv2.VideoCapture(video_path)

# change the image property
# use `$ v4l2-ctl --list-formats-ext` to check the available format!
# install first (if not yet): `$ sudo apt install v4l-utils`
cap.set(cv2.CAP_PROP_FRAME_WIDTH, published_weight)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, published_height)

if _enable_cv_out:
	cv2.namedWindow(window_title, cv2.WND_PROP_FULLSCREEN)
	# cv2.resizeWindow("Image", 1920, 1080)  # Enter your size
	cv2.resizeWindow(window_title, 800, 550)  # Enter your size
_frame_id = 0

# Extra information (to be tagged into the frame)
int_drone_id = encrypt_str("1")  # contains 1 extra slot
extra_len = 8  # contains 1 extra slot; another one slot is from `tagged_data_len` variable

# create an empty array
bw_usage_header = ['FrameID', 'Uncompressed', 'Compressed']
bw_usage = []
bw_usage_compressed = []

try:
	while cap.isOpened():
		_frame_id += 1

		# generate encrypted frame_id
		eframe_id = encrypt_str(str(_frame_id))

		# ret = a boolean return value from getting the frame, frame = the current frame being projected in the video
		try:
			ret, frame = cap.read()

			# do it ONCE
			# detect width and height
			if cam_weight is None:
				cam_height, cam_weight, _ = frame.shape

			img_size, ext = get_img_fsize_in_float(frame.nbytes)
			L.warning("[frame-{}] ## Image Size: {} {}".format(_frame_id, img_size, ext))
			# print(" ## Image Size Bytes:", fsizeb(frame.nbytes))
			L.warning("[frame-{}] ## Initial image SHAPE: {}".format(_frame_id, frame.shape))

			t0_decoding = time.time()
			# resize the frame; Default VGA (640 x 480) for Laptop camera
			if cam_weight != FULLHD_WIDTH and args.resize:
				frame = cv2.resize(frame, (FULLHD_WIDTH, FULLHD_HEIGHT))

			L.warning("[frame-{}] ## Final image SHAPE: {}".format(_frame_id, frame.shape))
			# compress image
			# NEW encoding method
			encoder_format = None  # disable custom encoder
			itype = 4  # new type specially for compressed image
			t0_img_compression = time.time()
			_, compressed_img = cv2.imencode('.jpg', frame, encode_param)
			compressed_img_len, _ = compressed_img.shape
			t1_img_compression = (time.time() - t0_img_compression) * 1000
			t1_img_compression = round(t1_img_compression, 3)
			L.warning(('[frame-{}][%s] Latency Image Compression (%.3f ms) '.format(_frame_id) % (
				datetime.now().strftime("%H:%M:%S"), t1_img_compression)))
			tagged_data_len = compressed_img_len + extra_len  # `tagged_data_len` itself contains 1 extra slot

			# create t0
			t0_array = str(time.time()).split(".")  # contains 2 extra slots

			# generate img compression latency
			img_compr_lat_arr = str(t1_img_compression).split(".")  # contains 2 extra slots

			# generate encrypted frame_id
			eframe_id = encrypt_str(str(_frame_id))

			# vertically tag this frame with an extra inforamtion
			t0_tag_extraction = time.time()
			tagged_info = [
				[int_drone_id],
				[int(t0_array[0])],
				[int(t0_array[1])],
				[eframe_id],
				[int(img_compr_lat_arr[0])],
				[int(img_compr_lat_arr[1])],
				[extra_len],
				[tagged_data_len],
			]
			val = np.vstack([compressed_img, tagged_info])
			t1_tag_extraction = (time.time() - t0_tag_extraction) * 1000
			L.warning(('[frame-{}][%s] Latency Image Taging (%.3f ms) '.format(_frame_id) % (datetime.now().strftime("%H:%M:%S"), t1_tag_extraction)))

			img_size_compressed, ext = get_img_fsize_in_float(val.nbytes)
			L.warning("[frame-{}] ## Image Size COMPRESSED + TAGGED: {} {}".format(_frame_id, img_size_compressed, ext))
			bw_usage.append([_frame_id, img_size, img_size_compressed])

			# publish data
			z_svc.publish(
				_val=val,
				_itype=itype,
			)

			if _enable_cv_out:
				cv2.imshow(window_title, frame)
			print()
		except Exception as e:
			print("No more frame to show: `{}`".format(e))
			break

		if _enable_cv_out:
			if cv2.waitKey(1) & 0xFF == ord('q'):
				break

# when stopped, start saving the CSV files
except KeyboardInterrupt:
	L.warning("[STOPPING] Start storing CSV Files")
	# if file exist, delete it first!
	if os.path.isfile(CSV_FILE_PATH):
		os.remove(CSV_FILE_PATH)

	# # open the file in the write mode
	with open(CSV_FILE_PATH, 'w', encoding='UTF8', newline='') as f:
		# create the csv writer
		writer = csv.writer(f)

		# write the header
		writer.writerow(bw_usage_header)

		# write a row to the csv file
		writer.writerows(bw_usage)

if _enable_cv_out:
	# The following frees up resources and closes all windows
	cap.release()
	cv2.destroyAllWindows()
#########################

# closing Zenoh publisher & session
z_svc.close_connection(publisher)
